Rewrite the step-17 in-code documentation.

bangerth opened the pull request:
Also some adjustments to the results section. In particular much
more focus on the limitations of this program, as well as trying
to give a better overview of how one has to think about parallel
programs.

In reference to #1365.

drwells commented on examples/step-17/doc/results.dox:
```diff
@@ -9,15 +9,15 @@ most basic way to run MPI programs is using a command line like
-that the exact command line syntax varies.
+(If you work on a cluster,
+there is typically a step in between where you need to set up a job script,
+submit the script to a scheduler that then executes the script whenever it
```
I had some trouble reading this sentence; the third clause is missing something. How about

```
If you work on a cluster, then there is typically a step in between where you
need to set up a job script and submit the script to a scheduler. The scheduler
will execute the script whenever it can allocate 32 unused processors for your
job.
```
instead?

drwells commented on examples/step-17/step-17.cc:
```diff
@@ -241,19 +251,20 @@ namespace Step17
+  // <code>this_mpi_process</code> is determined after creation of
+  // pcout, we cannot set the condition through the constructor,
+  // i.e. by <code>pcout(std::cout, this_mpi_process==0)</code>, but
+  // set the condition separately.
```
Is there any reason why we cannot initialize `this_mpi_process` at the top of the constructor (and also move things around in the declaration) and then use

```cpp
pcout(std::cout, this_mpi_process == 0)
```

to get rid of this discussion?

bangerth replied:
Yes, valid point. The discussion came from a time when we didn't have `Utilities::MPI::this_mpi_process()` and had to call `MPI_Commrank` which did not return the rank but instead put it into a variable; at that time, we couldn't initialize `this_mpi_process` either.

I'll change this.

drwells commented on examples/step-17/step-17.cc:
```diff
@@ -359,48 +416,55 @@ namespace Step17
+  // the local contributions into the global matrix or right hand side
+  // vector, we have to transfer these entries to the process that
+  // owns these elements. Fortunately, we don't have to do this by
+  // hand, PETSc does all this for us by caching these elements
```
This is not particularly important, but I believe this sentence should use a semicolon before PETSc and not a comma.

bangerth replied:
I'll do a colon instead. I assume that also works ;-)

drwells commented on examples/step-17/step-17.cc:
```diff
@@ -282,58 +293,97 @@ namespace Step17
+  // variable that indicates how many degrees of freedom the current
+  // process is responsible for. (This will, in general, be less than
+  // <code>fe.dofs_per_cell</code> times the number of cells the
+  // current process owns because some degrees of freedom live on
```
This argument seems strange because, unless we are using DG, won't `dofs_per_cell * n_active_cells` always overestimate since most `dof`s belong to more than one cell?

bangerth replied:
Yes, good point. Will reword.

bangerth replied:
The point I wanted to make is that DoFs are not necessarily owned by the owner of a cell it lives on because it may live on interfaces between subdomains. That was a rather poor way of describing this. Will fix.

drwells commented on examples/step-17/step-17.cc:
```diff
@@ -282,58 +293,97 @@ namespace Step17
+  // degrees of freedom that live on each cell, whether it is one that
+  // the current process owns or not. This can not scale to large
+  // problems because eventually just storing on every process the
+  // entire mesh and everything that is associated with it, will
```
Another grammatical nitpick: I believe there should be a comma after `entire mesh`.

drwells commented:
Wow, that was a lot of work just to review the changes. This is certainly a big improvement.

bangerth commented:
Yes, it's a lot of work to write the documentation of tutorial programs -- that's why we don't have more :-)

I've been going through some of them over the years and update the style in which it is written. In the early years, we wrote commentary for most every individual line, but I've come to find that that makes programs difficult to read because you don't see the forest for the trees. The newer style is to comment whole code blocks instead, by writing longer comments for whole sequences of lines. You can still see the difference if you look at, for example, http://dealii.org/developer/doxygen/deal.II/step_3.html#CommProg (old style) and http://dealii.org/developer/doxygen/deal.II/step_6.html#CommProg (mostly new style). But there is much more to do in this regard. I think I've also learned better what is important to explain and what may not -- the tutorials are full of stuff nobody cares about :-(

I think I've addressed all of your comments in the version that's now here.

drwells commented:
I found the philosophical parts of the tutorials very useful when I was still learning how to use the library so the effort is very much appreciated.

I have a two more small remarks:
1. In `results.dox` there are instructions for the user to 'in the Makefile[,] change the reference to the 2d libraries to their 3d counterparts'. Would you please remove that sentence?
2. Since this is a major update I think it would be appropriate to update the authorship information in the header.

bangerth commented:
> I found the philosophical parts of the tutorials very useful when I was still
> learning how to use the library so the effort is very much appreciated.

Yes. We need more of those. The pieces that show *how* are important, but the 
pieces that discuss *why* are also, and these have traditionally been missing.

> I have a two more small remarks:
> 1. In |results.dox| there are instructions for the user to 'in the Makefile[,]
> change the reference to the 2d libraries to their 3d counterparts'. Would you
> please remove that sentence?

Good catch!

> 2. Since this is a major update I think it would be appropriate to update the
> authorship information in the header.
Done.

drwells commented on examples/step-17/step-17.cc:
```diff
@@ -282,58 +290,97 @@ namespace Step17
+  // The final step of this initial setup is that we get ourselves a
+  // variable that indicates how many degrees of freedom the current
+  // process is responsible for. (Note that a degree of freedom is not
+  // necessarily owned by the process that owns a cells just because
```
should be 'owns a cell'

drwells commented:
Otherwise this is good to go. :+1:

kronbichler commented on examples/step-17/doc/intro.dox:
```diff
@@ -94,7 +94,11 @@ For larger problems, having to store the <i>entire</i> mesh on every processor
+numerous occasions where, in the course of discussing how a function of this
+program works, we will comment on the fact that it will not scale to large
+problems and why not. All of these issues will be addressed in step-18 and
+in particular step-40, which scales to very large number of processes.
```
number -> numbers (or 'a very large number')

kronbichler commented on examples/step-17/doc/results.dox:
```diff
@@ -9,15 +9,15 @@ most basic way to run MPI programs is using a command line like
   mpirun -np 32 ./step-17
 @endcode
-to run the step-17 executable with 32 processors.
+to run the step-17 executable with 32 processors. 
```
remove trailing whitespace

kronbichler commented on examples/step-17/step-17.cc:
```diff
@@ -116,23 +117,54 @@ namespace Step17
+
+    // Then we have two variables that tell us where in the parallel
+    // world we are. The first of the following variables,
+    // <code>n_mpi_processes</code> tells us how many MPI processes
```
I think you need a comma before 'tells'

kronbichler commented on examples/step-17/step-17.cc:
```diff
@@ -282,58 +290,97 @@ namespace Step17
+  // GridTools::partition_triangulation() function that does this at a
+  // much higher level of programming.
+  //
+  // @note As mentioned in the introduction, we can avoid this manual
```
I think 'can' should be replaced by 'could' (because it fits the 'used' in the next line)

bangerth commented:
I think I addressed all comments people mentioned. Thanks a lot for the careful reading!
